node/src/components/consensus/highway_core/state.rs

```diff
@@ -705,6 +709,149 @@ impl<C: Context> State<C> {
     pub(crate) fn is_empty(&self) -> bool {
         self.units.is_empty()
     }
+
+    /// Validates whether `wunit` violates the Limited Naïveté Criterion (LNC).
+    /// Returns index of the first equivocator that was cited naively in violation of the LNC, or
+    /// `None` if the LNC is satisfied.
+    fn validate_lnc(&self, wunit: &WireUnit<C>) -> Option<ValidatorIndex> {
+        wunit
+            .panorama
+            .enumerate()
+            .filter(|(_, obs)| obs.is_faulty())
+            .map(|(i, _)| i)
+            .find(|eq_idx| !self.satisfies_lnc_for(wunit, *eq_idx))
+    }
+
+    /// Returns `true` if there is at most one fork by the validator `eq_idx` that is cited naively
+    /// by `wunit` or earlier units by the same creator.
+    fn satisfies_lnc_for(&self, wunit: &WireUnit<C>, eq_idx: ValidatorIndex) -> bool {
+        // Find all forks by eq_idx that are cited naively by wunit itself.
+        // * If it's more than one, return false: the LNC is violated.
+        // * If it's none, return true: If the LNC were violated, it would be because of two naive
+        //   citations by wunit.creator's earlier units. So the latest of those earlier units would
+        //   already be violating the LNC itself, and thus would not have been added to the state.
+        // * Otherwise store the unique naively cited fork in naive_by_wunit.
+        let mut opt_naive_by_wunit = None;
+        {
+            // Returns true if any endorsed (according to wunit) unit cites the given unit.
+            let seen_by_endorsed =
+                |hash| wunit.endorsed.iter().any(|e_hash| self.sees(e_hash, hash));
+
+            // Iterate over all units cited by wunit.
+            let mut to_visit: Vec<_> = wunit.panorama.iter_correct_hashes().collect();
+            // This set is a filter so that units don't get added to to_visit twice.
+            let mut added_to_to_visit: HashSet<_> = to_visit.iter().cloned().collect();
```

Comment:
Why not just `visited`?

Reply:
Because they haven't been visited yet. They only have been added to `to_visit`.